app/src/main/java/com/oelrun/teta/movies/MoviesFragment.kt

```diff
@@ -74,34 +71,70 @@ class MoviesFragment: Fragment() {
             resources.getDimension(R.dimen.margin_main).toInt(),
             resources.getDimension(R.dimen.item_movie_margin_bottom).toInt()))
 
-        if(moviesData.isEmpty()) {
-            messageItem.visibility = View.VISIBLE
-        } else {
-            messageItem.visibility = View.GONE
-            adapterMovies.addHeaderAndSubmitList(moviesData) {
-                Handler(Looper.getMainLooper()).post {
-                    listMovies.invalidateItemDecorations()
+        val swipeContainer = view.findViewById<SwipeRefreshLayout>(R.id.swipe_container_movies)
+        swipeContainer.setOnRefreshListener {
+            moviesViewModel.loadMovies(true)
+        }
+
+        lifecycleScope.launch {
+            moviesViewModel.moviesData.collect { data ->
+                if(data.isEmpty()) {
+                    messageItem.visibility = View.VISIBLE
+                } else {
+                    messageItem.visibility = View.GONE
+                    adapterMovies.addHeaderAndSubmitList(data) {
+                        Handler(Looper.getMainLooper()).post {
+                            listMovies.invalidateItemDecorations()
+                        }
+                    }
+                    if(firstItemMovie != -1) {
+                        listMovies.smoothScrollToPosition(firstItemMovie)
+                    }
                 }
             }
-            listMovies.smoothScrollToPosition(firstItemMovie)
         }
 
-        listGenres = view.findViewById(R.id.list_genres)
+        lifecycleScope.launch {
+            moviesViewModel.isRefreshing.collect {
+                if(!it) {
+                    swipeContainer.isRefreshing = false
+                }
+            }
+        }
 
-        val data = GenresDataSource().getGenres()
+        lifecycleScope.launch {
+            moviesViewModel.errorMessage.collect {
```

Comment:
nit:      moviesViewModel.errorMessage.collect { message ->
                message?.let {
  Toast.makeText(view.context, it, Toast.LENGTH_SHORT).show()
                    moviesViewModel.errorMessageShown()
}
           }